Add exposing metrics via HTTP endpoint (#901)
memgraph/docs: docs/reference-guide/exposing-system-metrics.md

---
id: exposing-system-metrics
title: Exposing system metrics (Enterprise)
sidebar_label: Exposing system metrics
---

In production systems, monitoring of applications is crucial, and that includes databases as well. 
Memgraph allows tracking information about transactions, query latencies, snapshot recovery latencies, 
triggers, bolt messages, indexes, streams, and many more using an HTTP server.

Exposing metrics is a Memgraph Enterprise feature and therefore needs a valid Memgraph Enterprise license key.
After successfully entering the license key, Memgraph needs to be restarted in order to start the metrics HTTP server.

## Configuring the HTTP endpoint

The default address and port for the metrics server is `0.0.0.0:9091`, and can be configured using [configuration flags](/reference-guide/configuration.md)
`--metrics-address` and `--metrics-port`. If you need help changing the configuration follow [the how-to guide](/how-to-guides/config-logs.md).

## System metrics

> In Memgraph, there are 3 types of metrics that measure different parts of the system in order to ensure proper
> monitoring has been done. Those 3 metrics are:

Reviewer: ```suggestion
All system metrics measuring different parts of the system can be divided into three different types:
```

---
id: exposing-system-metrics
title: Exposing system metrics (Enterprise)
sidebar_label: Exposing system metrics
---

> In production systems, monitoring of applications is crucial, and that includes databases as well. Since Memgraph v2.8,
> an HTTP server has been added in order to expose system metrics, giving developers ability to track information about transactions,
> query latencies, snapshot recovery latencies, triggers, bolt messages, indexes, streams, and many more.

Reviewer: ```suggestion
In production systems, monitoring of applications is crucial, and that includes databases as well. Memgraph allows tracking information about transactions, query latencies, snapshot recovery latencies, triggers, bolt messages, indexes, streams, and many more.
```

the docs are versioned so this will appear only in version 2.8 indicating that the feature became available then

## Configuring the HTTP endpoint

> 
> Exposing metrics is a Memgraph Enterprise feature and therefore needs a valid Memgraph Enterprise license key.
> After successfully entering the license key, Memgraph would need to be restarted in order to start the metrics HTTP
> server in the following run.

Reviewer: ```suggestion
```

> The default address and port for the metrics server is `0.0.0.0:9091`, and can be configured with configuration flags
> `--metrics-address` and `--metrics-port`. For more info on the configuration, check the [list of the configuration variables](/reference-guide/configuration.md).

Reviewer: ```suggestion
The default address and port for the metrics server is `0.0.0.0:9091`, and can be configured using [configuration flags](/reference-guide/configuration.md) `--metrics-address` and `--metrics-port`. If you need help changing the configuration follow [the how-to guide](/how-to-guides/config-logs.md).
```

## Types of system metrics

> In Memgraph, there are 3 types of metrics that measure different parts of the system in order to ensure proper
> monitoring has been done. Those 3 metrics are:

Reviewer: ```suggestion
In Memgraph, there are 3 types of metrics that measure different parts of the system in order to ensure proper
monitoring. Those 3 metrics are:
```

> 
> ## Types of system metrics
> 
> In Memgraph, there are 3 types of metrics that measure different parts of the system in order to ensure proper
> monitoring has been done. Those 3 metrics are:
> - **Gauge** - a single value of some variable in the system (e.g. memory usage)
> - **Counter (uint64_t)** - a value that can be incremented or decremented (e.g. number of active transactions in the system)
> - **Histogram (uint64_t)** - distribution of measured values (e.g. certain percentile of query latency on N measured queries)
> 
> 

Reviewer: ```suggestion
```

Author: why do I need to delete everything?

Reviewer: I've repositioned it, but now i don't know where that suggestion disappeared!

> ## List of system metrics

Reviewer: ```suggestion
## System metrics
```

### General metrics

 | Name           | Type             | Description                                                 |
 | -------------- | ---------------- | ----------------------------------------------------------- |
 | vertex_count   | Gauge (uint64_t) | Number of nodes stored in the system.                       |
 | edge_count     | Gauge (uint64_t) | Number of relationships stored in the system.               |
 | average_degree | Gauge (double)   | Average number of relationships of a single node.           |
 | memory_usage   | Gauge (uint64_t) | Amount of RAM used reported by the OS (in bytes).           |

>  | disk_usage     | Gauge (uint64_t) | Amount of disk space used by the data directory (in bytes). |

Reviewer: ```suggestion
 | disk_usage     | Gauge (uint64_t) | Amount of disk space used by the [data directory](/reference-guide/backup.md) (in bytes). |
```

### Index metrics

 | Name                       | Type    | Description                                            |
 | -------------------------- | ------- | ------------------------------------------------------ |
 | ActiveLabelIndices         | Counter | Number of active label indices in the system.          |
 | ActiveLabelPropertyIndices | Counter | Number of active label property indices in the system. |

### Operator metrics

> :::info
> 
> For more info on operators, check [Inspecting queries](/reference-guide/optimizing-queries/inspecting-queries.md).
> 
> :::

Reviewer: ```suggestion
Before a Cypher query is executed, it is converted into an internal form suitable for execution, known as a plan. A plan is a tree-like data structure describing a pipeline of operations that will be performed on the database in order to yield the results for a given query. Every node within a plan is known as [a logical operator](/memgraph/reference-guide/inspecting-queries#operators) and describes a particular operation.
```

 | UnwindOperator                      | Counter | Number of times Unwind operator was used.                      |
 | DistinctOperator                    | Counter | Number of times Distinct operator was used.                    |
 | UnionOperator                       | Counter | Number of times Union operator was used.                       |
 | CartesianOperator                   | Counter | Number of times Cartesian operator was used.                   |
 | CallProcedureOperator               | Counter | Number of times CallProcedureOperator operator was used.       |
 | ForeachOperator                     | Counter | Number of times Foreach operator was used.                     |
 | EvaluatePatternFilterOperator       | Counter | Number of times EvaluatePatternFilter operator was used.       |
 | ApplyOperator                       | Counter | Number of times Apply operator was used.                       |

### Query metrics

 | Name                         | Type      | Description                                               |
 | ---------------------------- | --------- | --------------------------------------------------------- |

>  | QueryExecutionLatency_us_50p | Histogram | Query execution latency in microseconds (50th percentile) |
>  | QueryExecutionLatency_us_90p | Histogram | Query execution latency in microseconds (90th percentile) |
>  | QueryExecutionLatency_us_99p | Histogram | Query execution latency in microseconds (99th percentile) |

Reviewer: ```suggestion
 | QueryExecutionLatency_us_50p | Histogram | Query execution latency in microseconds (50th percentile). |
 | QueryExecutionLatency_us_90p | Histogram | Query execution latency in microseconds (90th percentile). |
 | QueryExecutionLatency_us_99p | Histogram | Query execution latency in microseconds (99th percentile). |
```

 | -------------- | ------- | -------------------------------------- |
 | ReadQuery      | Counter | Number of read-only queries executed.  |
 | WriteQuery     | Counter | Number of write-only queries executed. |
 | ReadWriteQuery | Counter | Number of read-write queries executed. |

### Session metrics

 | Name                    | Type    | Description                             |
 | ----------------------- | ------- | --------------------------------------- |
 | ActiveSessions          | Counter | Number of active connections.           |
 | ActiveBoltSessions      | Counter | Number of active Bolt connections.      |
 | ActiveTCPSessions       | Counter | Number of active TCP connections.       |
 | ActiveSSLSessions       | Counter | Number of active SSL connections.       |
 | ActiveWebSocketSessions | Counter | Number of active websocket connections. |
 | BoltMessages            | Counter | Number of Bolt messages sent.           |

### Snapshot metrics

 | Name                           | Type      | Description                                                 |
 | ------------------------------ | --------- | ----------------------------------------------------------- |

>  | SnapshotCreationLatency_us_50p | Histogram | Snapshot creation latency in microseconds (50th percentile) |
>  | SnapshotCreationLatency_us_90p | Histogram | Snapshot creation latency in microseconds (90th percentile) |
>  | SnapshotCreationLatency_us_99p | Histogram | Snapshot creation latency in microseconds (99th percentile) |
>  | SnapshotRecoveryLatency_us_50p | Histogram | Snapshot recovery latency in microseconds (50th percentile) |
>  | SnapshotRecoveryLatency_us_90p | Histogram | Snapshot recovery latency in microseconds (90th percentile) |
>  | SnapshotRecoveryLatency_us_99p | Histogram | Snapshot recovery latency in microseconds (99th percentile) |

Reviewer: ```suggestion
 | SnapshotCreationLatency_us_50p | Histogram | Snapshot creation latency in microseconds (50th percentile). |
 | SnapshotCreationLatency_us_90p | Histogram | Snapshot creation latency in microseconds (90th percentile). |
 | SnapshotCreationLatency_us_99p | Histogram | Snapshot creation latency in microseconds (99th percentile). |
 | SnapshotRecoveryLatency_us_50p | Histogram | Snapshot recovery latency in microseconds (50th percentile). |
 | SnapshotRecoveryLatency_us_90p | Histogram | Snapshot recovery latency in microseconds (90th percentile). |
 | SnapshotRecoveryLatency_us_99p | Histogram | Snapshot recovery latency in microseconds (99th percentile). |
```

 | StreamsCreated   | Counter | Number of streams created.            |
 | MessagesConsumed | Counter | Number of consumed streamed messages. |

### Transaction metrics

 | Name                   | Type    | Description                                                                     |
 | ---------------------- | ------- | ------------------------------------------------------------------------------- |
 | ActiveTransactions     | Counter | Number of active transactions.                                                  |
 | CommitedTransactions   | Counter | Number of committed transactions.                                               |
 | RollbackedTransactions | Counter | Number of rollbacked transactions.                                              |
 | FailedQuery            | Counter | Number of times executing a query failed (either during parse time or runtime). |

### Trigger metrics

 | Name             | Type    | Description                  |
 | ---------------- | ------- | ---------------------------- |
 | TriggersCreated  | Counter | Number of Triggers created.  |
 | TriggersExecuted | Counter | Number of Triggers executed. |

## Example response

> If we don't have any modifying configurations, by sending a GET request to `localhost:9091` in the 
> local Memgraph build, we will get response constructed as follows.

Reviewer: ```suggestion
If there aren't any modifying configurations, sending a GET request to `localhost:9091` in the 
local Memgraph build will result in a response similar to the one below: 
```
